### PR TITLE
Nattrak integration

### PR DESCRIPTION
**Pull request by AndyTWF**

### CODE REVIEW

File: src/datablock/DatablockFunctions.h [hunk 1]
```diff
@@ -5,5 +5,6 @@ namespace UKControllerPlugin {
 
         int ConvertAltitudeToFlightLevel(int altitude);
         int ConvertFlightLevelToAltitude(int flightLevel);
+        int NormaliseFlightLevelFromString(std::string level);
```
Comment (AxonC) [inline src/datablock/DatablockFunctions.h:8]:
```suggestion
        int NormaliseFlightLevelFromString(const std::string & level);

```

File: src/oceanic/Clearance.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+            // Extra information
+            std::string extra;
+
+            bool operator==(const Clearance& compare) const
+            {
+                return this->callsign == compare.callsign;
+            }
+
+            void operator=(const Clearance& compare)
```
Comment (AxonC) [inline src/oceanic/Clearance.h:46]: Is this not the default behaviour the compiler produces anyway?

File: src/oceanic/Clearance.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                this->entryTime = compare.entryTime;
+                this->clearanceIssued = compare.clearanceIssued;
+                this->extra = compare.extra;
+            }
+
+            const std::string CLEARANCE_STATUS_CLEARED = "CLEARED";
+            const std::string CLEARANCE_STATUS_PENDING = "PENDING";
+        } Clearance;
+    } // namespace Oceanic
+
```
Comment (AxonC) [inline src/oceanic/Clearance.h:63]: Rogue space :D

File: src/oceanic/OceanicEventHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+                if (!clearanceData.is_array()) {
+                    LogWarning("Unable to decode oceanic clearances from Nattrak, JSON is not array.");
+                    return;
+                }
+
+                // Loop the clearances and update local data
+                this->clearances.clear();
+                for (
+                    auto clearance = clearanceData.cbegin();
```
Comment (AxonC) [inline src/oceanic/OceanicEventHandler.cpp:50]: Does nlohmann offer a range iterator e.g `for (const auto a& : clearanceData)`?

File: src/oceanic/OceanicEventHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                    clearance != clearanceData.cend();
+                    ++clearance
+                ) {
+                    if (!NattrakClearanceValid(*clearance)) {
+                        LogWarning("Invalid clearance received from Nattrak");
+                        continue;
+                    }
+
+                    clearances.insert(
```
Comment (AxonC) [inline src/oceanic/OceanicEventHandler.cpp:59]:
```suggestion
                    this->clearances.insert(
```
Just so it's a bit clearer with similar variables in the function.

**Review by AxonC — CHANGES_REQUESTED** [review]

File: src/datablock/DatablockFunctions.h (hunk 1 above)
Comment (AndyTWF) [reply src/datablock/DatablockFunctions.h:8]: Done!

File: src/oceanic/Clearance.h (hunk 2 above)
Comment (AndyTWF) [reply src/oceanic/Clearance.h:46]: No unfortunately not - for a structs it needs the equality operator

File: src/oceanic/Clearance.h (hunk 3 above)
Comment (AndyTWF) [reply src/oceanic/Clearance.h:63]: Done!

File: src/oceanic/OceanicEventHandler.cpp (hunk 5 above)
Comment (AndyTWF) [reply src/oceanic/OceanicEventHandler.cpp:59]: Done!

File: src/oceanic/OceanicEventHandler.cpp (hunk 4 above)
Comment (AndyTWF) [reply src/oceanic/OceanicEventHandler.cpp:50]: Sadly not

File: src/datablock/DatablockFunctions.h (hunk 1 above)
Comment (AndyTWF) [reply src/datablock/DatablockFunctions.h:8]: A const reference doesn't work here due to the use of `std::transform` in the function itself

**Review by AxonC — APPROVED** [review]

**Pull request merged**